%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Beamer Presentation
% LaTeX Template
% Version 1.0 (10/11/12)
%
% This template has been downloaded from:
% http://www.LaTeXTemplates.com
%
% License:
% CC BY-NC-SA 3.0 (http://creativecommons.org/licenses/by-nc-sa/3.0/)
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%----------------------------------------------------------------------------------------
%	PACKAGES AND THEMEShttps://www.overleaf.com/project/5dbfadb88b2560000102ab44
%----------------------------------------------------------------------------------------

\documentclass{beamer}

\mode<presentation> {

% The Beamer class comes with a number of default slide themes
% which change the colors and layouts of slides. Below this is a list
% of all the themes, uncomment each in turn to see what they look like.

%\usetheme{default}
\usetheme{AnnArbor}
%\usetheme{Antibes}
%\usetheme{Bergen}
%\usetheme{Berkeley}
%\usetheme{Berlin}
%\usetheme{CambridgeUS}
%\usetheme{Copenhagen}
%\usetheme{Darmstadt}
%\usetheme{Dresden}
%\usetheme{Frankfurt}
%\usetheme{Goettingen}
%\usetheme{Hannover}
%\usetheme{Ilmenau}
%\usetheme{JuanLesPins}
%\usetheme{Luebeck}
%\usetheme{Madrid}
%\usetheme{Malmoe}
%\usetheme{Marburg}
%\usetheme{Montpellier}
%\usetheme{PaloAlto}
%\usetheme{Pittsburgh}
%\usetheme{Rochester}
%\usetheme{Singapore}
%\usetheme{Szeged}
%\usetheme{Warsaw}

% As well as themes, the Beamer class has a number of color themes
% for any slide theme. Uncomment each of these in turn to see how it
% changes the colors of your current slide theme.

%\usecolortheme{albatross}
%\usecolortheme{beaver}
%\usecolortheme{beetle}
%\usecolortheme{crane}
%\usecolortheme{dolphin}
%\usecolortheme{dove}
%\usecolortheme{fly}
%\usecolortheme{lily}
%\usecolortheme{orchid}
%\usecolortheme{rose}
%\usecolortheme{seagull}
%\usecolortheme{seahorse}
%\usecolortheme{whale}


%\usetheme{Boadilla}
\usecolortheme{wolverine}

%\setbeamertemplate{footline} % To remove the footer line in all slides uncomment this line
\setbeamertemplate{footline}[page number] % To replace the footer line in all slides with a simple slide count uncomment this line

\setbeamertemplate{navigation symbols}{} % To remove the navigation symbols from the bottom of all slides uncomment this line
}

\setbeamertemplate{bibliography item}{\insertbiblabel}


\usepackage{graphicx} % Allows including images
\usepackage{booktabs} % Allows the use of \toprule, \midrule and \bottomrule in tables
%\usepackage {tikz}
\usepackage{tkz-graph}
\usepackage{bibentry}


\usepackage[
backend=biber,
style=numeric,
sorting=ynt
]{biblatex}
 
\documentclass[xcolor=table]{beamer} 

\bibliography{bibfile}

\graphicspath{ {images/} }

\GraphInit[vstyle = Shade]
\tikzset{
  LabelStyle/.style = { rectangle, rounded corners, draw,
                        minimum width = 2em, fill = yellow!50,
                        text = red, font = \bfseries },
  VertexStyle/.append style = { inner sep=5pt,
                                font = \normalsize\bfseries},
  EdgeStyle/.append style = {->, bend left} }
\usetikzlibrary {positioning}
%\usepackage {xcolor}
\definecolor {processblue}{cmyk}{0.96,0,0,0}

% show sections only
\setcounter{tocdepth}{1}
% show subsections
%\setcounter{tocdepth}{2}
% show subsubsections
%\setcounter{tocdepth}{3}
\AtBeginSection[]
{
    \begin{frame}
        \frametitle{Table of Contents}
        \tableofcontents[currentsection]
    \end{frame}
}

%\AtBeginSubsection[]
%{
%    \begin{frame}
%        \frametitle{Table of Contents}
%        \tableofcontents[currentsection,currentsubsection]
%    \end{frame}
%}

%----------------------------------------------------------------------------------------
%	TITLE PAGE
%----------------------------------------------------------------------------------------

\title[caDDS]{caDDS: Community Assisted Distributed Database for Sequences} 

%%%%%%%%%%%%%%%%%%%%%%%%%%% INFO %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\author{Azcarraga, Araya} % Your name
\institute[Department of Computer Science, University of the Philippine - Diliman] % Your institution as it will appear on the bottom of every slide, may be shorthand to save space
{
University of the Philippine - Diliman\\ % Your institution for the title page
\medskip
}
\date{\today} % Date, can be changed to a custom date


%----------------------------------------------------------------------------------------
%	PRESENTATION SLIDES
%----------------------------------------------------------------------------------------


\begin{document}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%% TITLE %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}
\titlepage % Print the title page as the first slide
\end{frame}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%% OVERVIEW %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}
\frametitle{Overview} % Table of contents slide, comment this block out to remove it
\tableofcontents % Throughout your presentation, if you choose to use \section{} and \subsection{} commands, these will automatically be printed on this slide as an overview of your presentation
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%% INTRO + RRL %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}{Abstract}
  Genomic data the past few years have grown exponentially with the technology barely keeping up. The problem faced by genome researchers is the large data set and difficulty transferring the files. Previous distributed databases are either not meant for genomic data, or difficult to replicate. This paper lays the groundwork for a distributed database that is designed to easily scale to accommodate bigger storage, and also reduces the data speed bottleneck faced by centralized databases. The proposed system has a master node and data nodes. Data nodes should theoretically increase data transfer speeds even as the number of users increase. (to be updated based on conclusion) While current databases(e.g. NCBI) have more data, more users, and a live platform. The proposed system is aimed towards a smaller community, with more frequent and localized data, which the data storage can easily be expanded as the need arises. A recommendation is to implement the system and make the code public for the use of other communities.
\end{frame}
\section{Introduction}
\subsection{Genomic Concepts}
\begin{frame}{Genes}
  \begin{itemize}   
    \item all organisms have genes
    \item Genes are made of a certain molecular structure called DNA which correspond to information denoted by: A,G,C,T
    \item genes are instructions and blueprints for life
\end{itemize}
\end{frame}

\begin{frame}{Human Genes}
  \begin{itemize}   
    \item There is a lot of information in one human cell
    \item One human cell's entire genes laid out is 2 meters in length \cite{ency_sci_tech} and has $3.2x10^9$ base pairs \cite{introgenomics}
    \item Human body has a lot of information, that when laid out spans the earth to the moon six thousand times
\end{itemize}
\end{frame}


\begin{frame}{Genomics}
  \begin{block} {\textbf{Definition of Genomics}}
    Genomics is the study of all of a person's genes (the genome), including interactions of those genes with each other and with the person's environment. \cite{genomics-definition}
  \end{block}
  \begin{itemize}   
    \item \textbf{Genomics}: study of whole sets of genes and their interactions \cite{campbell}
    \item More complex organisms have more complex genes (many parts and interactions)
    \item Studying these complex organisms needs a lot of storage
    \item The current online portal mainly used to study this is NCBI \cite{campbell}
  \end{itemize}
\end{frame}

\begin{frame}{Sequencing}
  \begin{itemize}   
    \item Reading genes from cells is sequencing
    \item Replications (coverage) \& Fragmentation of DNA is necessary to read
    \item Every read stores the many fragments
    \item Each sequence read is redundant and uses a lot of space
  \end{itemize}
\end{frame}

\begin{frame}{Human Genome Project Successes}
  \begin{itemize}   
    \item 2 years after, the project led to a whole genome project for chimpanzees (human's closest relative) \cite{campbell}
    \item These led to many whole genome analysis of organisms (around 4,300 in 2013) \cite{campbell}
    \item Growing amount of research and data from genome research.
  \end{itemize}
\end{frame}

\begin{frame}{Human Genome Project}
  \begin{itemize}   
    \item Finished in 2003, project wanted to read the entire human genome
    \item Didn't have the mature tools at the time
    \item Big cost, many institutions, 10 years \cite{introgenomics}
  \end{itemize}
\end{frame}

\begin{frame}{Next Generation Sequencing}
  \begin{itemize}   
    \item New sequencing technology to read faster
    \item Faster reads corresponds to more data
  \end{itemize}
\end{frame}

\begin{frame}{Moore's Law}
  \begin{block}{\textbf{Moore's law}} 
    states that the number of transistors on a chip doubles every two years.\cite{kurose}
  \end{block}
  \begin{itemize}   
    \item Prediction of the speed new technology (faster processing power and better storage) is developed
    \item But moore's law is too slow for the speed of genomic data increase
    \item
  \end{itemize}
\end{frame}

\begin{frame}{Growing problem of storage}
  \begin{itemize}   
    \item Each read\cite{seqtorr} has 60x coverage, around 200gb each
    \item with NGS, more genomic researches, more files needed to be saved and transferred
  \end{itemize}
\end{frame}

\subsection{Database Concepts}
\begin{frame}{Database}
  \begin{itemize}   
    \item Technology for data storage
    \item Technology needed to be developed to resolve the growing storage concern
  \end{itemize}
\end{frame}

\begin{frame}{Database Definition}
A database or a DBMS (DatBase Management System) \cite{Silberschatz2010} is defined as:
\begin{itemize}
    \item Collection of interrelated data
    \item Set of programs to access the data
    \item An environment that is both convenient and efficient to use
\end{itemize}
\end{frame}

\begin{frame}{Database Principles}
  \begin{itemize}   
    \item consistency
    \item availability
    \item partition tolerance
  \end{itemize}
\end{frame}

\begin{frame}{Database Types}
  \begin{itemize}   
    \item Centralized
    \item Distributed
    \item Distributed will be focused since it has the ability to speed up transfer speeds
  \end{itemize}
\end{frame}

\begin{frame}{Application Architecture}
  \begin{itemize}   
    \item Client Server
    \item P2P
    \item Hybrid
    \item Hybrid architecture will be used because the Client-Server has bottlenects, and the P2P has no centralized management
  \end{itemize}
\end{frame}

\section{Review of Related technology}

\begin{frame}{SeqTorr}
  \begin{itemize}   
    \item idea
    \item pros
    \item cons
  \end{itemize}
\end{frame}

\begin{frame}{BioTorrents}
  \begin{itemize}   
    \item idea
    \item pros
    \item cons
  \end{itemize}
\end{frame}

\begin{frame}{PeerDB}
  \begin{itemize}   
    \item idea
    \item pros
    \item cons
  \end{itemize}
\end{frame}

\begin{frame}{base system: NCBI}
  \begin{itemize}   
    \item why popular
    \item idea
    \item idea
  \end{itemize}
\end{frame}

\begin{frame}{base system: NCBI}
  \begin{itemize}   
    \item pros
    \item cons
    \item cons
  \end{itemize}
\end{frame}

\begin{frame}{Table}
% Bio-related -> biological
% What's the difference between biological vs general vs FASTA
% What does general mean?
% What does biological mean? -> explain pa to
% May 3 types of data ...
   \begin{table}[]
        \begin{tabular}{|l|l|l|l|l|}
        \hline
        System      & CS/P2P/H & Available? & DL/UL speed & Type of Data \\ \hline
        BioTorrents & P2P           & No         & Depends                    & Biological  \\ \hline
        SeqTorr     & H             & No         & Fast (PH server)           & FASTA        \\ \hline
        PeerDB      & P2P           & No         & Depends                    & General      \\ \hline
        NCBI        & CS            & Yes        & Slow (US server)           & Biological  \\ \hline
        \end{tabular}
    \end{table} 
\end{frame}

\section{Problem}
\begin{frame}{Problem}
  \begin{itemize}   
    \item classical databases
    \item distributed databases
  \end{itemize}
\end{frame}

\section{Objectives}
\begin{frame}{Objectives}
  \begin{itemize}   
    \item make a dist database
    \item community assisted
    \item main APIs
    \item time complexity
    \item pros and cons
  \end{itemize}
\end{frame}

\section{Scope}
\begin{frame}{Scope}
  \begin{itemize}   
    \item no this
    \item and
    \item that
  \end{itemize}
\end{frame}


\section{Theoretical Framework}
\begin{frame}{Network}
  \begin{itemize}   
    \item five levels
    \item need to focus on Level 4 and 5 only
    \item 
  \end{itemize}
\end{frame}

\subsection{System Summary}
\begin{frame}{System}
  \begin{itemize}   
    \item system is this
    \item distributed
    \item community assisted
  \end{itemize}
\end{frame}

\begin{frame}{Summary}
  \begin{itemize}   
    \item 
    \item 
    \item 
  \end{itemize}
\end{frame}

\subsection{Community}
\begin{frame}{Community}
  \begin{itemize}   
    \item good for community
    \item easy to expand storage
    \item 
  \end{itemize}
\end{frame}

\subsection{Comparisons}
\begin{frame}{Speed Comparisons}
  \begin{itemize}   
    \item simple time complexity chart
    \item faster si caDDS ofc 
    \item 
  \end{itemize}
\end{frame}


\begin{frame}{Final Comparisons}
  \begin{itemize}   
    \item caDDS is scalable, easy for community, has docu hehe
    \item NCBI is one, stable, widely used
    \item 
  \end{itemize}
\end{frame}


\begin{frame}{Recommendations}
  \begin{itemize}   
    \item implement
    \item security
    \item public code
  \end{itemize}
\end{frame}

\begin{frame}
\Huge{\centerline{The End}}
\small{\centerline{Thank you for coming}}
\end{frame}

\section{Bibliography}

\begin{frame}[allowframebreaks]
\printbibliography[heading=none]
\end{frame}


\end{document}

